plugins/intrinsics/src/main/java/org/qbicc/plugin/intrinsics/core/CoreIntrinsics.java

```diff
@@ -1017,9 +1022,50 @@ static void registerOrgQbiccRuntimeValuesIntrinsics(final CompilationContext ctx
         StaticIntrinsic isAlwaysFalse = (builder, target, arguments) -> literalOf(ctxt, arguments.get(0) instanceof BooleanLiteral && ((BooleanLiteral) arguments.get(0)).booleanValue());
         intrinsics.registerIntrinsic(valsDesc, "isAlwaysFalse", boolBoolDesc, isAlwaysFalse);
 
-        // todo: compareAndSwap*
 
-        // getAndSet*
+        // compareAndSwap*
+        class CompareAndSwapIntrinsic implements StaticIntrinsic {
+            private final MemoryAtomicityMode successMode;
+            private final MemoryAtomicityMode failureMode;
+            private final boolean isVolatile;
+            CompareAndSwapIntrinsic(final MemoryAtomicityMode successMode, final MemoryAtomicityMode failureMode,
+                                    boolean isVolatile) {
+                this.successMode = successMode;
+                this.failureMode = failureMode;
+                this.isVolatile = isVolatile;
+            }
+
+            @Override
+            public Value emitIntrinsic(BasicBlockBuilder builder, MethodElement element, List<Value> arguments) {
+                ValueHandle target = getTarget(ctxt, builder, arguments.get(0));
+                if (target == null) {
+                    return ctxt.getLiteralFactory().literalOf(false);
+                }
+                Value expect = arguments.get(1);
+                Value update = arguments.get(2);
+                Value result = builder.cmpAndSwap(target, expect, update, successMode, failureMode, isVolatile);
+                Value resultValue = builder.extractMember(result, ((CmpAndSwap)result).getResultValueType());
+                /* set was successful when expected value is returned */
+                return builder.isEq(resultValue, expect);
+            }
+        }
+
+        StaticIntrinsic compareAndSwapVolatile = new CompareAndSwapIntrinsic(MemoryAtomicityMode.MONOTONIC, MemoryAtomicityMode.MONOTONIC, true);
+        StaticIntrinsic compareAndSwapAcquire = new CompareAndSwapIntrinsic(MemoryAtomicityMode.ACQUIRE, MemoryAtomicityMode.ACQUIRE, false);
```

Comment:
Is ACQUIRE correct for both success and failure?  I'm having some trouble mapping this to the equivalent Unsafe operations but I would have expected the same failure mode for this case and the RELEASE case below.

Reply:
I'm not sure about this one but agree it makes sense to be consistent.